token-erc-20/chaincode-go/chaincode/token_contract.go

```diff
@@ -84,11 +83,104 @@ func (s *SmartContract) Mint(ctx contractapi.TransactionContextInterface, amount
 	totalSupply += amount
 	ctx.GetStub().PutState(totalSupplyKey, []byte(strconv.Itoa(totalSupply)))
 
+	// Emit the Transfer event
+	transferEvent := map[string]string{"from": "0x0", "to": minter, "value": strconv.Itoa(amount)}
+	transferEventJSON, err := json.Marshal(transferEvent)
+	if err != nil {
+		return fmt.Errorf("failed to obtain JSON encoding: %v", err)
+	}
+	err = ctx.GetStub().SetEvent("Transfer", transferEventJSON)
+	if err != nil {
+		return fmt.Errorf("event failed to register: %v", err)
+	}
+
+	log.Printf("minter account %s balance updated from %d to %d", minter, currentBalance, updatedBalance)
+
 	return nil
 }
 
-// Transfer transfers tokens from client account to recipient account.
-// recipient account must be a valid clientID as returned by the ClientID() function.
+// Burn redeems tokens the minter's account balance
+// This function triggers a Transfer event
+func (s *SmartContract) Burn(ctx contractapi.TransactionContextInterface, amount int) error {
+
+	// Check minter authorization - this sample assumes Org1 is the central banker with privilege to burn new tokens
+	clientMSPID, err := ctx.GetClientIdentity().GetMSPID()
+	if err != nil {
+		return fmt.Errorf("failed to get MSPID: %v", err)
+	}
+	if clientMSPID != "Org1MSP" {
+		return fmt.Errorf("client is not authorized to mint new tokens")
+	}
+
+	// Get ID of submitting client identity
+	minter, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get client id: %v", err)
+	}
+
+	if amount <= 0 {
+		return fmt.Errorf("burn amount must be a positive integer")
+	}
+
+	currentBalanceBytes, err := ctx.GetStub().GetState(minter)
+	if err != nil {
+		return fmt.Errorf("failed to read minter account %s from world state: %v", minter, err)
+	}
+
+	var currentBalance int
+
+	// Check if minter current balance exists
+	if currentBalanceBytes == nil {
+		return fmt.Errorf("The balance does not exist")
+	} else {
+		currentBalance, _ = strconv.Atoi(string(currentBalanceBytes)) // Error handling not needed since Itoa() was used when setting the account balance, guaranteeing it was an integer.
+	}
+
+	updatedBalance := currentBalance - amount
+
+	err = ctx.GetStub().PutState(minter, []byte(strconv.Itoa(updatedBalance)))
+	if err != nil {
+		return err
+	}
+
+	// Update the totalSupply
+	totalSupplyBytes, err := ctx.GetStub().GetState(totalSupplyKey)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve total token supply: %v", err)
+	}
+
+	var totalSupply int
+
+	// If no tokens have been minted, throw error
+	if totalSupplyBytes == nil {
+		return fmt.Errorf("totalSupply does not exist")
+	}
+
+	totalSupply, _ = strconv.Atoi(string(totalSupplyBytes)) // Error handling not needed since Itoa() was used when setting the totalSupply, guaranteeing it was an integer.
```

Comment:
Nit: related to the above comment
```suggestion
	totalSupply, _ := strconv.Atoi(string(totalSupplyBytes)) // Error handling not needed since Itoa() was used when setting the totalSupply, guaranteeing it was an integer.
```